services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -0,0 +1,326 @@
+import io.grpc.reflection.v1alpha.ServiceResponse;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * Provides a reflection service for Protobuf services (including the reflection service itself).
+ */
+public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
+    implements NotifyOnServerBuild {
+  private Server server;
+  private Set<String> serviceNames;
+  private Map<String, FileDescriptor> fileDescriptorsByName;
+  private Map<String, FileDescriptor> fileDescriptorsBySymbol;
+  private Map<String, Map<Integer, FileDescriptor>> fileDescriptorsByExtensionAndNumber;
+  private Boolean mapsInitialized = false;
+
+  /**
+   * Retrieves the services registered to the server at build time, and stores those services with a
+   * {@link ProtoFileDescriptorSupplier}.
+   */
+  public void notifyOnBuild(Server server) {
+    this.server = server;
+  }
+
+  private void processExtension(FieldDescriptor extension, FileDescriptor fd) {
+    String extensionName = extension.getContainingType().getFullName();
+    int extensionNumber = extension.getNumber();
+    if (fileDescriptorsByExtensionAndNumber.containsKey(extensionName)) {
+      fileDescriptorsByExtensionAndNumber.get(extensionName).put(extensionNumber, fd);
+    } else {
+      Map<Integer, FileDescriptor> extensionMap = new HashMap<Integer, FileDescriptor>();
+      extensionMap.put(extensionNumber, fd);
+      fileDescriptorsByExtensionAndNumber.put(extensionName, extensionMap);
+    }
+  }
+
+  private void processService(ServiceDescriptor service, FileDescriptor fd) {
+    fileDescriptorsBySymbol.put(service.getFullName(), fd);
+    for (MethodDescriptor method : service.getMethods()) {
+      fileDescriptorsBySymbol.put(method.getFullName(), fd);
+    }
+  }
+
+  private void processType(Descriptor type, FileDescriptor fd) {
+    fileDescriptorsBySymbol.put(type.getFullName(), fd);
+    for (FieldDescriptor extension : type.getExtensions()) {
+      processExtension(extension, fd);
+    }
+    for (Descriptor nestedType : type.getNestedTypes()) {
+      processType(nestedType, fd);
+    }
+  }
+
+  private void processFileDescriptor(FileDescriptor fd) {
+    fileDescriptorsByName.put(fd.getName(), fd);
+    for (ServiceDescriptor service : fd.getServices()) {
+      processService(service, fd);
+    }
+    for (Descriptor type : fd.getMessageTypes()) {
+      processType(type, fd);
+    }
+    for (FieldDescriptor extension : fd.getExtensions()) {
+      processExtension(extension, fd);
+    }
+  }
+
+  /**
+   * Processes the service file descriptors. This prepares a list of the service names and indexes
+   * the file descriptors by name, symbol, and extension.
+   */
+  private synchronized void initFileDescriptorMaps() {
```

Comment:
I guess my question is more, why is this one method synchronized, unlike the rest of the methods in this class.